packages/shaders/src/shader-mount.ts

```diff
@@ -665,3 +676,58 @@ function isSafari() {
   const ua = navigator.userAgent.toLowerCase();
   return ua.includes('safari') && !ua.includes('chrome') && !ua.includes('android');
 }
+
+// Zoom level can be estimated comparing the browser's outerWidth and the viewport width.
+// It's nowhere near perfect because it's affected by the presence of browser sidebars,
+// like a vertical web inspector or Arc's sidebar. Also, both outerWidth and innerWidth
+// are integers, which would almost never give us a perfect ratio at face values.
+//
+// Still, this is pretty accurate in the vast majority of cases.
+//
+// Note 1:
+// Avoid innerWidth, use visualViewport.width instead.
+// - innerWidth is affected by pinch zoom in Safari, but not other browsers.
+//   visualViewport.width works consistently in all browsers.
+// - innerWidth is rounded to integer, but not visualViewport.width.
+// - visualViewport.width is affected by hard scrollbars, so they need to be added manually
+//
+// Note 2:
+// Opening a sidebar in Safari like web inspector or bookmarks will throw off the zoom
+// level detection and result in a larger target resolution. Not a concern in real-world usage
+// with Safari, but we'd rather not try to detect zoom levels with other browsers
+// (e.g. Arc always has a sidebar, which affects outerWidth vs visualViewport.width).
+function bestGuessBrowserZoom() {
+  const viewportScale = visualViewport?.scale ?? 1;
+  const viewportWidth = visualViewport?.width ?? window.innerWidth;
+  const scrollbarWidth = window.innerWidth - document.documentElement.clientWidth;
+  const innerWidth = viewportScale * viewportWidth + scrollbarWidth;
+
+  // outerWidth and innerWidth are always integers so we won't often get the original zoom ratio
+  // E.g. given a 125% zoom, outerWidth = 1657, innerWidth = 1325, 1657 / 1325 = 1.2505660377
+  // We check for common zoom levels and return the closest one if found.
+
+  const ratio = outerWidth / innerWidth;
+  const zoomPercentageRounded = Math.round(100 * ratio);
+
+  // All zoom levels divisible by 5%
+  if (zoomPercentageRounded % 5 === 0) {
+    return zoomPercentageRounded / 100;
+  }
+
+  // 33% zoom
+  if (zoomPercentageRounded === 33) {
+    return 1 / 3;
+  }
+
+  // 67% zoom
+  if (zoomPercentageRounded === 67) {
+    return 2 / 3;
+  }
+
+  // 133% zoom
```

Comment:
is this the furthest we have to go? could 167 and beyond zoom happen?

Reply:
Haven't seen 167% in the wild, even 133% isn't common (only Firefox does it)